governance/default.go

```diff
@@ -665,6 +666,14 @@ func getGovernanceCacheKey(num uint64) common.GovernanceCacheKey {
 
 // Store new governance data on DB. This updates Governance cache too.
 func (g *Governance) WriteGovernance(num uint64, data GovernanceSet, delta GovernanceSet) error {
+	g.idxCacheLock.RLock()
+	idx := make([]uint64, 0, len(g.idxCache))
+	copy(idx, g.idxCache)
```

Comment:
I do not seem to use idx. Why do I copy?

Reply:
Just to get the `len(idx)`?

Reply:
@nogang @ehnuje It is to get `len(idx)` and `idx[len(idx)-1]`. I think copying can be removed if the checks are done inside the lock.

Reply:
@nogang @ehnuje I think the code can be changed like this
```go
	g.idxCacheLock.RLock()
	defer g.idxCacheLock.RUnlock()

	if len(g.idxCache) > 0 && num <= g.idxCache[len(g.idxCache)-1] {
		return nil
	}
```
But in this case, the lock on idxCache have to last until the writing is done if it is for actually new governance information.

Reply:
@andybclee `defer g.idxCacheLock.RUnlock()` can make deadlock with `.addGovernanceCache(num, new)` on line 685.

Reply:
@KimKyungup @ehnuje Yes. It makes a problem. I made a separate function which just check the condition and to be called by `WriteGovernance` and `addGovernanceCache`. PTAL.